modules/manifold/src/utils/default-states.js

```diff
@@ -0,0 +1,279 @@
+        })
+      );
+    })
+  );
+};
+
+/**
+ * compute segmentFilter from a single field definition
+ * @param {Object} field - contains {name, type, tableFieldIndex, role, dataType, domain}
+ * @returns {Array<Object>} segmentFilters, each contains {name, key, type, value}
+ */
+export const defaultSegmentFiltersFromFieldDef = field => {
+  const {name, tableFieldIndex} = field;
+  const type = defaultFilterTypeFromFieldDef(field);
+  const filterVals = defaultFilterValueFromFieldDef(field);
+  return filterVals.map(filterVal => ({
+    name,
+    key: tableFieldIndex - 1,
+    type,
+    value: filterVal,
+  }));
+};
+
+/**
+ * @param {Object} field - contains {name, type, tableFieldIndex, role, dataType, domain}
+ * @returns {String} filterType - one of FILTER_TYPE
+ */
+const defaultFilterTypeFromFieldDef = field => {
+  const {type} = field;
+  switch (type) {
+    case FEATURE_TYPE.NUMERICAL:
+      return FILTER_TYPE.RANGE;
+    case FEATURE_TYPE.CATEGORICAL:
+      return FILTER_TYPE.INCLUDE;
+    default:
+      return FEATURE_TYPE.FUNC;
+  }
+};
+
+/**
+ * @param {Object} field - contains {name, type, tableFieldIndex, role, dataType, domain}
+ * @returns {Array<Object>} segmentFilters, each contains {name, key, type, value}
+ */
+const defaultFilterValueFromFieldDef = field => {
+  const {role, type, domain} = field;
+  assert(
+    Array.isArray(domain) && domain.length > 1,
+    'domain length must be larger than 1'
+  );
+  let midPoint;
+  switch (type) {
+    case FEATURE_TYPE.NUMERICAL:
+      // for selecting over/under prediction
+      // todo: score columns don't have domain field right now. To add.
+      if (
+        role === FIELD_ROLE.SCORE &&
+        domain[0] < 0 &&
+        domain[domain.length - 1] > 0
+      ) {
+        midPoint = 0;
+      } else {
+        midPoint = (domain[0] + domain[domain.length - 1]) / 2;
+      }
+      return [[domain[0], midPoint], [midPoint, domain[domain.length - 1]]];
+
+    case FEATURE_TYPE.CATEGORICAL:
+      return [[domain[0]], [domain[1]]];
+
+    default:
+      return [[], []];
+  }
+};
+
+//todo: consolidate with `isValidFilterVals` in modules/manifold/src/components/ui/segment-panels/utils.js
+/**
+ * given a filter and the corresponding field definition, determine whether the filter is valid
+ * @param  {Object} filter - contains {name, key, type, value}
+ * @param  {object} field - contains {name, tableFieldIndex, type, dataType, role, domain}
+ * @return {boolean}
+ */
+export const isValidSegmentFilterFromFieldDef = (filter, field) => {
+  const {type: filterType, value} = filter;
+  const {type: featureType, domain} = field;
+
+  if (featureType === FEATURE_TYPE.CATEGORICAL) {
+    assert(
+      domain && domain.length && domain.length,
+      '`field.domain` must be an array'
+    );
+    return (
+      filterType === FILTER_TYPE.INCLUDE &&
+      // filter must consain a subset of column domain
+      value.length < domain.length &&
```

Comment:
can `value.length === domain.length`?

Reply:
No, we assume a filter should contain only a subset of original data.